mlrun/api/db/sqldb/db.py

```diff
@@ -665,6 +674,204 @@ def delete_project(self, session, name: str):
     def list_projects(self, session, owner=None):
         return self._query(session, Project, owner=owner)
 
+    def get_feature_set(
+        self, session, project: str, name: str, tag: str = None, hash_key: str = None
+    ) -> schemas.FeatureSet:
+        query = self._query(session, FeatureSet, name=name, project=project)
+        computed_tag = tag or "latest"
+        tag_fs_uid = None
+        if not tag and hash_key:
+            uid = hash_key
+        else:
+            tag_fs_uid = self._resolve_tag_function_uid(
+                session, FeatureSet, project, name, computed_tag
+            )
+            if tag_fs_uid is None:
+                raise mlrun.errors.MLRunNotFoundError(
+                    f"Feature-set tag not found {project}/{name}:{tag}"
+                )
+            uid = tag_fs_uid
+        if uid:
+            query = query.filter(FeatureSet.uid == uid)
+        obj = query.one_or_none()
+        if obj:
+            db_fs = self._transform_feature_set_model_to_scheme(obj)
+
+            # If connected to a tag add it to metadata
+            if tag_fs_uid:
+                db_fs.metadata.tag = computed_tag
+            return db_fs
+        else:
+            raise mlrun.errors.MLRunNotFoundError(
+                f"Feature-set not found {project}/{name}"
+            )
+
+    # TODO - IMPLEMENT QUERY BY LABEL!!!!!
+    def list_feature_sets(
+        self,
+        session,
+        project: str,
+        name: str = None,
+        tag: str = None,
+        state: str = None,
+        entities: List[str] = None,
+        features: List[str] = None,
+        labels: List[str] = None,
+    ) -> List[schemas.FeatureSet]:
+        # Find object IDs by tag, project and fs-name (which is a like query)
+        tag = tag or "latest"
+        tag_query = self._query(session, FeatureSet.Tag, project=project, name=tag)
+        if name:
+            tag_query = tag_query.filter(FeatureSet.Tag.obj_name.ilike(f"%{name}%"))
+        obj_uids = {row.obj_id: row.name for row in tag_query}
+
+        # Query exact fields
+        query = self._query(session, FeatureSet, project=project, state=state)
+
+        if name is not None:
+            query = query.filter(FeatureSet.name.ilike(f"%{name}%"))
+        if obj_uids:
+            query = query.filter(FeatureSet.id.in_(obj_uids.keys()))
+        if entities:
+            ent_alias = aliased(Feature)
+            query = query.join(FeatureSet.entities.of_type(ent_alias)).filter(
+                ent_alias.name.in_(entities)
+            )
+        if features:
+            feat_alias = aliased(Feature)
+            query = query.join(FeatureSet.features.of_type(feat_alias)).filter(
+                feat_alias.name.in_(features)
+            )
+
+        feature_sets = []
+        for db_fs in query:
+            fs = self._transform_feature_set_model_to_scheme(db_fs)
+            fs.metadata.tag = obj_uids[db_fs.id]
```

Comment:
But then if tag is None (query all tags) then the returned feature-sets will not have any tag associated with them. What I do here is I fill the actual tag for that specific object from the result of the query over the tag table.

Reply:
Mmm you're right, but, remember that one object can have many tags, so the mapping you did here is also not complete.
You can take a look on what I did in `list_functions`, TBH it can probably be done better performance-wise (instead of querying tags table per uid, querying once), but you should have something like that here, note that there is also where I handle the untagged unversioned case (discussed in another comment)

Reply:
OK. I think I got you now. I've implemented something similar, but as you mentioned I'm using a single query over the tags table and use my lookup table to store the resulting tags. Should be fixed now.